project_euler/problem_074/sol2.py

```diff
@@ -86,25 +94,27 @@ def solution(chain_length: int = 60, number_limit: int = 1000000) -> int:
     for i in range(1, number_limit + 1):
 
         # The temporary list will contain the elements of the chain
-        chain_list = [i]
+        chain_set = {i}
+        len_chain_set = 1
+        last_chain_element = i
 
         # The new element of the chain
-        new_chain_element = factorial_sum(chain_list[-1])
+        new_chain_element = factorial_sum(last_chain_element)
 
         """ Stop computing the chain when you find a repeating item
             or the length it greater then the desired one.
         """
-        while not (new_chain_element in chain_list) and (
-            len(chain_list) <= chain_length
-        ):
-            chain_list += [new_chain_element]
+        while new_chain_element not in chain_set and len_chain_set <= chain_length:
+            chain_set.add(new_chain_element)
 
-            new_chain_element = factorial_sum(chain_list[-1])
+            len_chain_set += 1
+            last_chain_element = new_chain_element
+            new_chain_element = factorial_sum(last_chain_element)
 
-        """ If the while exited because the chain list contains the exact amount of elements
-            increase the counter
+        """ If the while exited because the chain list contains the exact amount
+            of elements increase the counter
         """
-        chain_counter += len(chain_list) == chain_length
+        chain_counter += len_chain_set == chain_length
```

Comment:
What does this expression do?
Is it adding the length to the counter? Why is it doing a comparison as well?

Reply:
The variable `chain_counter` which is the solution basically is increased by 1 if the length of the chain is same as the chain length we are looking for.
The right side checks equality and returns 1 (True) or 0 (False), which is the value with which the counter is increased.

Reply:
That would be quite confusing for a beginner. Please separate out the logic.

Reply:
I have changed it, just so I can merge this PR. If you see the diff, this was already there and the issue was for optimization, not refactoring.
Personally, I am all for improving the code quality along the way with fixing bugs and enhancements and would have welcomed it if it would have been mentioned in the first PR review itself and not when the PR is ready to be merged.

Honestly, the unrelated nit-picking really frustrated me.

And FYI, this is called branchless programming and it is reasonably efficient over if statements. Here's a link for more info - https://www.youtube.com/watch?v=bVJ-mWWL7cE , if you're curious.